### ISSUE TITLE
Document how to dynamically define a function when using a module

### ISSUE THREAD

**Issue by krader1961** [elves/elvish#1133]
I was never happy with my *~/.elvish/lib/cmd-duration.elv* module including this function definition:

```
fn now {
    use math

    if (has-external gdate) {
        # Presumably we're on a macOS system where I've installed the GNU
        # date utility.
        put (math:round (* 1000 (gdate +%s.%N)))
    } else {
        try {
            put (math:round (* 1000 (date +%s.%N)))
        } except e {
            # Must be a *BSD system which doesn't support %N and where we
            # don't have GNU date. So simply accept that a resolution of
            # seconds is the best we can do.
            put (math:round (* 1000 (date +%s)))
        }
    }
}
```

The problem is that it does potentially expensive tests every single time it is invoked. The `now` function in that module should instead be dynamically defined at "use" time resulting in a function that is optimal for the platform. That, however, is difficult because function definitions, like lambdas, are lexically scoped.

I improved the efficiency of that module by instead doing this:

```
fn now []{ }
if (has-external gdate) {
    now~ = []{ float64 (gdate '+%s.%N') }
} else {
    now~ = [] { float64 0 }
}
```

I've omitted irrelevant details from the above example (in particular the content of the "else" block). My point is that how to dynamically define a function in a module is not obvious. Either the technique I used should be documented or there should be a more straightforward mechanism for doing the same thing.

**Comment by hanche** [comment]
Perhaps worth noting, here is a different way to achieve the same result:

```
now~ = (
  if (has-external gdate) {
    put []{ float64 (gdate '+%s.%N') }
  } else {
    put [{ float64 0 }]
  }
)
```

**Comment by krader1961** [comment]
@hanche, That's an interesting approach to the problem; however, it's the same length as my solution and, I think, slightly less clear. Still, it hadn't occurred to me so well worth mentioning.

Note: My solution could be further optimized by making the default definition the implementation expected to be used the most often. That would eliminate one `now~ = {...}` assignment.

**Comment by krader1961** [comment]
> Either the technique I used should be documented or there should be a more straightforward mechanism for doing the same thing.

I can't help but think that explicit function definitions, via the `fn` keyword, should be scoped to the namespace rather than lexically scoped. Unlike anonymous lambdas. That difference in behavior has its own cognitive overhead. But I think the difference in behavior is easy to justify, document, and is more likely to comport with user's expectations. Note that the existing documentation regarding closures does not use an explicit `fn` defined lambda in any of its examples -- they only use anonymous lambdas.

**Comment by hanche** [comment]
> That's an interesting approach to the problem; however, it's the same length as my solution and, I think, slightly less clear. Still, it hadn't occurred to me so well worth mentioning.

Agreed, mostly. However, my solution has the advantage of being slightly more DRY, as there is no need to repeat the name of the function being defined.

**Comment by hanche** [comment]
> I can't help but think that explicit function definitions, via the `fn` keyword, should be scoped to the namespace rather than lexically scoped.

I am not sure what you mean here.

When I first came to elvish from es, I thought of `fn foo []{ }` as merely being syntactic sugar for `foo~ = []{ }`. And to some extent it is, except for inner function definitions, which are _always_ locally scoped. Perhaps you meant to abolish the exception, so that an inner `fn foo []{ }` would overwrite an outer `foo~`, if it exists, but otherwise, remain local to the function. If so, I can sympathise, but not sure I agree.

Or perhaps you mean that an inner `fn foo []{ }` should _always_ create a function name in the outer scope, in which case I strongly disagree. It would confuse the heck out of me, and possibly break some of my code. Using inner helper functions in a complicated function is a pretty common pattern. It may be true that one should try to avoid writing complicated functions, but not all complexity is best handled by writing and composing a myriad of tiny functions.

Somewhat related to this issue is #645, forcing you to declare an inner variable with `val` or `var`. Perhaps we need a similar mechanism for dealing with function definitions using `fn`. But I can't come up with a reasonable syntax.

> Note that the existing documentation regarding closures does not use an explicit `fn` defined lambda in any of its examples -- they only use anonymous lambdas.

True, but I think the [given example](https://elv.sh/ref/language.html#closure-semantics) would be better written using `fn`-defined functions:

```
fn make-adder {
  n = 0
  fn putter { put $n }
  fn adder { n = (+ $n 1) }
  put $putter~ $adder~
}
```

It is clearer, especially if the two exported functions were any longer, as they would be in any non-trivial example.

**Comment by krader1961** [comment]
> Or perhaps you mean that an inner fn foo []{ } should always create a function name in the outer scope

Yes.

> When I first came to elvish from es, I thought of fn foo []{ } as merely being syntactic sugar for foo~ = []{ }. And to some extent it is, except for inner function definitions, which are always locally scoped.

That isn't true. In particular `fn` traps the exception raised by `return` while an anonymous lambda does not. It's one of the reasons I created commit 3300c81c.

Your rewrite of the `make-addr` example is interesting. But isn't this version just as clear (at least to anyone who knows elvish well enough to recognize a lambda):

```
fn make-addr {
    n = 0
    putter_fn = { put $n }
    addr_fn = { n = (+ $n 1) }
    put $putter_fn $addr_fn
}
```

Having said all that I also think a `fn` definition in a block, rather than at the module level, should be lexically scoped rather than module scoped. Primarily because that's needed for closure semantics. Nonetheless, I wish there were a more straightforward method for defining module level functions inside a block rather than at the module top-level.

**Comment by hanche** [comment]
@krader1961 Oops, you're right. My bad – I forgot about the `return` capture bit. (I added a note my previous comment.)
